frontends/p4/sideEffects.cpp

```diff
@@ -310,40 +310,205 @@ class DismantleExpression : public Transform {
     const IR::Node* preorder(IR::LAnd* expression) override { return shortCircuit(expression); }
     const IR::Node* preorder(IR::LOr* expression) override { return shortCircuit(expression); }
 
-    // We don't want to compute the full read/write set here so we
-    // overapproximate it as follows: all declarations that occur in
-    // an expression.
-    // TODO: this could be made more precise, perhaps using LocationSets.
-    class ReadsWrites : public Inspector {
+    /// This class represents the path to a location.
+    /// Given a struct S { bit a; bit b; } and a variable S x;
+    /// a path can be x.a, or just x.  An array index is represented as a
+    /// number (encoded as a string) or as "*", denoting an unknown index.
+    struct LocationPath : public IHasDbPrint {
+        const IR::IDeclaration* root;
+        std::vector<cstring> path;
+
+        explicit LocationPath(const IR::IDeclaration* root): root(root) { CHECK_NULL(root); }
+
+        const LocationPath* append(cstring suffix) const {
+            auto result = new LocationPath(root);
+            result->path = path;
+            result->path.push_back(suffix);
+            return result;
+        }
+
+        /// True if this path is a prefix of other or the other way around
+        bool isPrefix(const LocationPath* other) const {
+            // Due to the structure of the P4 language, two distinct
+            // declarations can never alias.
+            if (root != other->root)
+                return false;
+            size_t len = std::min(path.size(), other->path.size());
+            for (size_t i = 0; i < len; i++) {
+                if (path.at(i) == "*" || other->path.at(i) == "*")
+                    continue;
+                if (path.at(i) != other->path.at(i))
+                    return false;
+            }
+            return true;
+        }
+
+        void dbprint(std::ostream& out) const override {
+            out << root->getName();
+            for (auto p : path)
+                out << "." << p;
+        }
+    };
+
+    /// We represent a set of location set as a set of LocationPath
+    /// objects.
+    class SetOfLocations : public IHasDbPrint {
      public:
-        std::set<const IR::IDeclaration*> decls;
-        ReferenceMap* refMap;
+        std::set<const LocationPath*> paths;
```

Comment:
Using sets of pointers can lead to non-deterministic behavior, as the order of iteration over the set depends on what the garbage collector does.  Since all the iterations here *should* be order independent, that's probably not an issue here, but that's what we have `ordered_set` and `ordered_map` for,